### ISSUE TITLE
Add pkg-config (.pc) file

### ISSUE THREAD

**Issue by haasn** [google/shaderc#392]
`pkg-config` is the standard way to find C libraries on Linux and related systems. It's typical of libraries that intend to be used on Linux to package a pkg-config file.

As far as I can tell, shaderc doesn't seem to include one.

**Comment by dneto0** [comment]
True, Shaderc doesn't have one.  So far most users either compile from source or get it from an SDK such as the Vulkan SDK or Android NDK.

Can you provide a reference to the file format, or a pull request to add one?

**Comment by haasn** [comment]
A pkg-config file is useful even when compiling from source, especially considering you seem to vary the library name based on whether or not it's shared or static (for whatever reason).

Here's an example of a pkg-config file:

`/usr/lib64/pkgconfig/flac.pc`:
```
prefix=/usr
exec_prefix=${prefix}
libdir=/usr/lib64
includedir=${prefix}/include

Name: FLAC
Description: Free Lossless Audio Codec Library
Version: 1.3.2
Requires.private: ogg
Libs: -L${libdir} -lFLAC
Libs.private: -lm
Cflags: -I${includedir}
```

You can find a general overview here: https://en.wikipedia.org/wiki/Pkg-config and a more in-depth documentation here: https://linux.die.net/man/1/pkg-config

**Comment by dneto0** [comment]
Note to self. The same thing was done for SPIRV-Tools via
https://github.com/KhronosGroup/SPIRV-Tools/pull/1103  and https://github.com/KhronosGroup/SPIRV-Tools/pull/1284

**Comment by polyzen** [comment]
Would this enable one to build Shaderc with system-wide installed build deps as opposed to using `third_party/`?

**Comment by haasn** [comment]
> Would this enable one to build Shaderc with system-wide installed build deps as opposed to using third_party/?

No, the issues are orthogonal. This would simply allow *others* to build against shaderc more easily, specifically also allowing users to do e.g. version checks in configure scripts.

As for building shaderc itself against non-vendored third party libraries, that just requires changing the shaderc build script slightly. You can actually find examples of this being done for e.g. the [gentoo ebuild](https://github.com/gentoo/gentoo/blob/master/media-libs/shaderc/shaderc-9999.ebuild), which refuses to use vendored libraries in favor of system wide ones.

**Comment by sl1pkn07** [comment]
any update about this?

greetings

**Comment by fjhenigman** [comment]
I threw something together, completely stolen from SPIRV-Tools.
https://github.com/google/shaderc/pull/581
The .pc files look reasonable to me but I haven't tried using them.  If someone could check them that would be great.
Though I think we need -lshaderc_util somewhere, in the same .pc file or a separate one, not sure which.
And do we want a .pc file for libshaderc_combined?
I think we will want a separate one for libshaderc_spvc.

**Comment by fjhenigman** [comment]
Duh, just realized there's one easy test:
PKG_CONFIG_PATH=. pkg-config --libs shaderc
Works.

**Comment by sl1pkn07** [comment]
if i understand corrct, shaderc.pc is for static libs and shaderc-shared.pc is for shared lisb. rigth?

if the static library if a set of 2 files, libshaderc and libshaderc-combined.a . is not should be add the `-lshaderc-combined` to the shaderc.pc file?

if is a separate lib. why libshasderc-combined not have the shared lib counterpart?

as my understand. separate naming can be avoided . because the unique difference (i mean in naming) between the static lib is the extension, `a` or `lib` for static and `.so` and .`dll` for shared. use other name schame is a mess

also. pkgfiles can be include info for shared and static libs. see example here https://stackoverflow.com/questions/27586503/how-to-use-pkg-config-to-link-a-library-statically

greetings

**Comment by fjhenigman** [comment]
Yes shaderc.pc is for the static version, shaderc-shared.pc for the shared/dynamic version.
I don't know why the the shared version has "-shared" in the name but will find out.  I can see how it's not good with pkgconfig.
The libshaderc-combined.a is a static lib with all dependent libs built in.  Maybe a shared version could be made, but it doesn't seem to be of much value.  I may as well make a shaderc-combined.pc too.
And I think we want shaderc_util.pc and reference it in the "Requires.private" line of shaderc.pc.

**Comment by sl1pkn07** [comment]
what happend if:

~~~patch
diff --git a/libshaderc/CMakeLists.txt b/libshaderc/CMakeLists.txt
index 9d07136..c6543d1 100644
--- a/libshaderc/CMakeLists.txt
+++ b/libshaderc/CMakeLists.txt
@@ -22,6 +22,7 @@ target_compile_definitions(shaderc_shared
     PUBLIC SHADERC_SHAREDLIB
 )
 set_target_properties(shaderc_shared PROPERTIES SOVERSION 1)
+set_target_properties(shaderc_shared PROPERTIES OUTPUT_NAME shaderc)
 
 if(SHADERC_ENABLE_INSTALL)
   install(
diff --git a/libshaderc_spvc/CMakeLists.txt b/libshaderc_spvc/CMakeLists.txt
index fae235e..af8d787 100644
--- a/libshaderc_spvc/CMakeLists.txt
+++ b/libshaderc_spvc/CMakeLists.txt
@@ -28,6 +28,7 @@ if (DISABLE_EXCEPTIONS)
 endif (DISABLE_EXCEPTIONS)
 
 set_target_properties(shaderc_spvc_shared PROPERTIES SOVERSION 1)
+set_target_properties(shaderc_spvc_shared PROPERTIES OUTPUT_NAME shaderc_spvc)
 
 if(SHADERC_ENABLE_INSTALL)
   install(
~~~

?

i've this this and the libshader_shared is renamed to libshared.so

~~~
┌─┤[$]|[sl1pkn07]|[sL1pKn07]|[~/aplicaciones/shaderc/build/esee/usr/local/lib64]|
└───╼  ldd *
libshaderc.a:
ldd: atención: no tiene permiso de ejecución para `./libshaderc.a'
        no es un ejecutable dinámico
libshaderc_combined.a:
ldd: atención: no tiene permiso de ejecución para `./libshaderc_combined.a'
        no es un ejecutable dinámico
libshaderc.so:
        linux-vdso.so.1 (0x00007fff0731b000)
        libglslang.so => /usr/lib/libglslang.so (0x00007f9424f12000)
        libOSDependent.so => /usr/lib/libOSDependent.so (0x00007f9424f0d000)
        libOGLCompiler.so => /usr/lib/libOGLCompiler.so (0x00007f9424f08000)
        libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f9424ee7000)
        libSPIRV.so => /usr/lib/libSPIRV.so (0x00007f9424762000)
        libSPIRV-Tools.so => /usr/lib/libSPIRV-Tools.so (0x00007f9424645000)
        libHLSL.so => /usr/lib/libHLSL.so (0x00007f94245a9000)
        libSPIRV-Tools-opt.so => /usr/lib/libSPIRV-Tools-opt.so (0x00007f9424340000)
        libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f94241b1000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f942402c000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f9424012000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f9423e4e000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x00007f942517b000)
        librt.so.1 => /usr/lib/librt.so.1 (0x00007f9423e42000)
libshaderc.so.1:
        linux-vdso.so.1 (0x00007ffdcb72d000)
        libglslang.so => /usr/lib/libglslang.so (0x00007fe5ecd95000)
        libOSDependent.so => /usr/lib/libOSDependent.so (0x00007fe5ecd90000)
        libOGLCompiler.so => /usr/lib/libOGLCompiler.so (0x00007fe5ecd8b000)
        libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007fe5ecd6a000)
        libSPIRV.so => /usr/lib/libSPIRV.so (0x00007fe5ec5e5000)
        libSPIRV-Tools.so => /usr/lib/libSPIRV-Tools.so (0x00007fe5ec4c8000)
        libHLSL.so => /usr/lib/libHLSL.so (0x00007fe5ec42c000)
        libSPIRV-Tools-opt.so => /usr/lib/libSPIRV-Tools-opt.so (0x00007fe5ec1c3000)
        libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007fe5ec034000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007fe5ebeaf000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007fe5ebe95000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007fe5ebcd1000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x00007fe5ecffe000)
        librt.so.1 => /usr/lib/librt.so.1 (0x00007fe5ebcc5000)
┌─┤[$]|[sl1pkn07]|[sL1pKn07]|[~/aplicaciones/shaderc/build/esee/usr/local/lib64]|
└───╼  file *
libshaderc.a:          current ar archive
libshaderc_combined.a: current ar archive
libshaderc.so:         symbolic link to libshaderc.so.1
libshaderc.so.1:       ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=eef47f6b426b251ecac7afe78015fb9e7836a67c, with debug_info, not stripped
~~~

greetings

**Comment by dcbaker** [comment]
Please don't build a "combined" or use inheritance in pkg-config, that's not normal. Shaderc is a bit strange because it provides both a static and a shared library at the same time, linux projects usually only provide one or the other and require the project to be rebuilt and installed twice if you want both. It would be preferable however of the `shaderc.pc` to provide the **shared** version, as that is what people on Linux generally want. In fact, you probably don't even need to provide a pkg-config file for the static version at all.

tl;dr It's fine to just provide a single `shaderc.pc` that provides the shared library, if someone wants to statically link shaderc they can just vendor/submodule it.

**Comment by fjhenigman** [comment]
We don't want to break any existing users, so we're not going to rename anything.
We also don't want to add libshaderc.so (while keeping libshaderc.a) because people who were getting the static lib with -lshaderc would start geting the shared lib since it takes precedence.
We can add shaderc.pc that will output -lshaderc_shared from the command "pkg-config shaderc"
For completeness (at the risk of adding confusion) we'll add shaderc_static.pc and shaderc_combined.pc too.
Hope this helps.

**Comment by knappador** [comment]
The impacts of this quality-of-life improvement will begin accruing almost immediately.

https://github.com/google/shaderc-rs/issues/44

I am implementing the system search fallback phase of fixing the issue and I have to for now just guess the path.

Anyone happen  to know where homebrew etc like to stash their shaderc static libs?

**Comment by fjhenigman** [comment]
Landed https://github.com/google/shaderc/commit/439a848c0c0e91b5f1b85059cd4b059274b9a93b so closing this.

**Comment by sl1pkn07** [comment]
THANKS!

greetings!

**Comment by eli-schwartz** [comment]
> Shaderc is a bit strange because it provides both a static and a shared library at the same time, linux projects usually only provide one or the other and require the project to be rebuilt and installed twice if you want both.

That's not even true at all, many projects compile both. autotools does this by default, unless you explicitly use --disable-static. meson has -Ddefault_library which supports "both" but defaults to shared, and leaves this in the control of the user. Projects which use bare makefiles offer separate targets for both, but usually share object files between the two and don't require running `make clean` in order to build both.

Many system distributions will choose to provide both, in order to cater to users wishing to build their own statically linked programs.

>  We also don't want to add libshaderc.so (while keeping libshaderc.a) because people who were getting the static lib with -lshaderc would start geting the shared lib since it takes precedence.

I would argue that they should have always used `-Wl,-static` (since that is "best practice" for people who care whether they link statically or dynamically), but I guess if they were being encouraged to not bother explicitly specifying how they wanted to link, then this would constitute a breaking change for them.

It is nevertheless true that using plain old `-lfoo` on the compiler command line is "supposed" to mean, "I have no bias regarding what type of library to link to and am merely requesting that the compiler should link to *something*".

**Comment by eebssk1** [comment]
I'v been building ffmpeg partial static. Now every time ffmpeg picks up the shared shaderc, I have to patch the config script to look for shaderc_static instead of plain shaderc.
